lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -431,6 +431,18 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
   auto Lookup = lookupUnqualified(DC, Name, Loc, LookupOptions);
 
   if (!Lookup) {
+    // Recover from not finding 'Self' in a nominal type, if we are
```

Comment:
That precedence is used currently so the protocol's dynamic `Self` parameter takes is used over this. I do agree that this is kind of problematic though — where would you suggest putting the check?